[feat] Add support for Tron Staking 2.0 (balance & history)

Philippoes opened the pull request:
### 📝 Description

Adds the following operation modes to enable Tron Stake 2.0 in Ledger Live (Primarily to use with the Ledger Wallet API).

- [`freeze -> FreezeBalanceV2Contract`]( https://developers.tron.network/reference/freezebalancev2-1) - To freeze/stake TRX and obtain voting rights (Tron Power) + Bandwidth or Energy
- [`unfreeze -> UnfreezeBalanceV2Contract`](https://developers.tron.network/reference/unfreezebalancev2-1) - To unfreeze/unstake TRX and release the corresponding amount of bandwidth or energy, and voting rights (Tron Power)
- [`unDelegateResource`](https://developers.tron.network/reference/undelegateresource-1) - For cases when a user has delegated resources and wants to unstake an amount larger than what's just "Frozen".
- [`withdrawExpireUnfreeze`](https://developers.tron.network/reference/withdrawexpireunfreeze) - To withdraw unFrozenV2 balances and get back the funds after the 14-day unfreezing period has passed. 
- `legacyUnfreeze`- To unfreeze legacy (Tron Stake 1.0) TRX, in case a user still has balance frozen with 1.0. 

...The above operations and associated balances (`frozen`, `delegatedFrozen`, `unfrozen`, `legacyFrozen`) are now part of the account balance. Meaning, that they will be part of the historical operations and asset balance on the accounts page. The PR includes minor adjustments to LLD and LLM to be able to display the operations and balance.




### ❓ Context

We're from [StakeKit](https://www.stakek.it/) and we've just released a Ledger Live app that offers users a unified non-custodial staking experience across a wide range of networks and protocols. We're opening this PR because we would like to offer users the possibility to participate in Tron 2.0 staking via our Ledger Live app. We will also open up another PR in the [wallet-api](https://github.com/LedgerHQ/wallet-api) repository that enables the new modes. FYI, the `DelegateResource` operation is out of scope as per our requirements. `unDelegateResource` has been added in case a user would like to fully unstake their funds through our UI and needs to undelegate the resources before unfreezing them.

### ✅ Checklist

Pull Requests must pass the CI and be code reviewed. Set as Draft if the PR is not ready.

- [x] `npx changeset` was attached.
- [x] **Covered by automatic tests.**
- [x] **Impact of the changes:** <!-- Please take some time to list the impact & what specific areas Quality Assurance (QA) should focus on -->
  - This PR impacts the portfolio balance and historical transactions of Tron accounts for both LLD and LLM.

---

### 🧐 Checklist for the PR Reviewers

<!-- Please do not edit this if you are the PR author -->

- [ ] **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- [ ] **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- [ ] **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- [ ] **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- [ ] **Any new dependencies** have been justified and documented.
- [ ] **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

haammar-ledger reviewed (commented):
Thanks a lot for your contribution.

A general question: since Tron staking v1 isn't active anymore, wouldn't it make sense to just replace v1 fields and functions with v2 equivalents?
- "un/freeze" could just be updated/replaced, no need for additional "un/freezev2"
- Not sure about legacy "vote" and "claimRewards", but if unused they should either be removed or updated with v2 equivalents

WDYT?

Philippoes commented:
> Thanks a lot for your contribution.
> 
> A general question: since Tron staking v1 isn't active anymore, wouldn't it make sense to just replace v1 fields and functions with v2 equivalents?
> 
> * "un/freeze" could just be updated/replaced, no need for additional "un/freezev2"
> * Not sure about legacy "vote" and "claimRewards", but if unused they should either be removed or updated with v2 equivalents
> 
> WDYT?

You're welcome.
 
Since users can technically still have frozen funds tied to Tron Stake 1.0 we'd like to keep support for unfreeze. We can for sure remove the freeze v1 operation as it's deprecated.

The "vote" and "claimRewards" operation modes are the same for Tron Stake 2.0, so no changes are required.

Philippoes commented:
@haammar-ledger After some thought, I realized it could be beneficial (For UX, translations, etc) to just replace the v1 with the v2 methods and add another operation for "LegacyUnfreeze". Pushing some changes...

haammar-ledger commented:
@Philippoes I can see you're still working on this PR, just let me know when this is ready and you want us to review again :+1: 
You can also set the PR to "draft" then set it as "ready to review" again later if you want.

Thanks!

Philippoes commented:
@haammar-ledger It is ready for review now. I just rebased the PR yesterday to resolve a conflict. 👍 

Thank you!

haammar-ledger commented on libs/ledger-live-common/src/families/tron/bridge/js.ts:
```diff
@@ -567,7 +619,7 @@ const getTransactionStatus = async (a: TronAccount, t: Transaction): Promise<Tra
     errors.recipient = new RecipientRequired();
   }
 
-  if (["send", "freeze", "unfreeze"].includes(mode)) {
```
Is the `freeze` case not handled anymore?

Philippoes replied:
Yeah, it seemed unnecessary since we don't use `recipient` in the new `FreezeTransactionData` and `UnFreezeTransactionData`.

haammar-ledger commented on apps/ledger-live-desktop/src/renderer/families/tron/AccountBalanceSummaryFooter.tsx:
```diff
@@ -63,14 +63,32 @@ const AccountBalanceSummaryFooter = ({ account }: Props) => {
+      .plus(BigNumber(frozenEnergyAmount || 0))
+      .plus(BigNumber(delegatedFrozenBandwidthAmount || 0))
+      .plus(BigNumber(delegatedFrozenEnergyAmount || 0))
+      .plus(BigNumber(unfrozenBandwidthAmount))
+      .plus(BigNumber(unfrozenEnergyAmount)),
```
Just for my understanding, `unfrozenBandwidthAmount` and `unfrozenEnergyAmount` are actually pat of the frozen amount because they are not available yet, until the unfreezing period is complete, is this correct?

Philippoes replied:
Correct. The unfrozen balances are still technically frozen until the user withdraws the funds. However, the Tron Power and resources will be released immediately upon calling `Unfreeze`.

haammar-ledger commented on apps/ledger-live-desktop/static/i18n/en/app.json:
```diff
@@ -2027,7 +2030,7 @@
         "amountLabel": "Amount to freeze",
         "available": "Available: {{amountAvailable}}",
-        "info": "Frozen assets cannot be sent for 3 days."
+        "info": "Frozen assets can be unfrozen immediately and accessible after the unfreezing period."
```
How long is the unfreezing period with Tron staking v2?

Philippoes replied:
14 days. Will add it to this phrase.

haammar-ledger replied:
@Philippoes You can add it on the mobile translations as well. Thanks.

Philippoes commented on libs/ledger-live-common/src/families/tron/types.ts:
```diff
@@ -101,18 +118,36 @@ export type SmartContractTransactionData = {
+  unfreeze_balance: number;
+};
+
+export type LegacyUnfreezeTransactionData = {
```
It's essentially the same as before. It might appear differently due to how Git is visualizing the changes.

Before:
```
export type UnfreezeTransactionData = {
  receiver_address?: string;
  owner_address: string;
  resource: TronResource | null | undefined;
};
```

After:
```
export type LegacyUnfreezeTransactionData = {
  receiver_address?: string;
  owner_address: string;
  resource: TronResource | null | undefined;
};
```

Philippoes commented on libs/ledger-live-common/src/families/tron/bridge/js.ts:
```diff
@@ -140,6 +146,15 @@ const signOperation: SignOperationFnSignature<Transaction> = ({ account, transac
+              case "unDelegateResource":
+                return unDelegateResourceTransaction(account, transaction);
+
+              case "legacyUnfreeze":
+                return legacyUnfreezeTronTransaction(account, transaction);
+
```
Correct!

Philippoes commented on libs/ledger-live-common/src/families/tron/utils.ts:
```diff
@@ -32,11 +32,15 @@ export const hexToAscii = (hex: string): string => Buffer.from(hex, "hex").toStr
   "VoteWitnessContract",
   "WithdrawBalanceContract",
   "ExchangeTransactionContract",
+  "FreezeBalanceV2Contract",
+  "UnfreezeBalanceV2Contract",
+  "WithdrawExpireUnfreezeContract",
+  "UnDelegateResourceContract",
+  "FreezeBalanceContract",
+  "UnfreezeBalanceContract",
```
Yes, the UnfreezeBalanceContract is the legacy unfreeze operation. IMO as long as the operation modes are mapped correctly to the respective TrongridTxType (and are aligned with [Tron transaction types](https://github.com/tronprotocol/java-tron/blob/develop/protocol/src/main/protos/core/Tron.proto#L343)) it should be alright. Kind of related to what we discussed earlier about replacing the v2 with the v1 operations.

Philippoes commented:
@haammar-ledger Just rebased the branch and fixed a small issue in the transaction status logic, and also updated the phrases in the mobile translations as suggested. Feel free to take a look!

haammar-ledger reviewed (commented):
Changes look ok :+1:  just a couple failures in the CI:
- https://github.com/LedgerHQ/ledger-live/actions/runs/7249706782/job/19748265503
  - `pnpm doc:ljs`
- https://github.com/LedgerHQ/ledger-live/actions/runs/7249706518/job/19748263171
  - `pnpm mobile lint` & fix found issue(s)

Philippoes commented:
@haammar-ledger Thanks! I've pushed a fix, and the CI should pass now. 👍

haammar-ledger commented:
@Philippoes One very last test failing :slightly_smiling_face: if you want to have a look

https://github.com/LedgerHQ/ledger-live/actions/runs/7277060700/job/19828256925#step:8:3665

Almost there!

Philippoes commented:
@haammar-ledger Thanks for pointing this out. The failing test is resolved, also added some integration tests for more coverage. Let's give the CI another go! 👍

haammar-ledger commented:
@Philippoes Congrats on passing the review and the CI tests!
We'll just do a last step of manual QA before merging, please bear with us, this should be done shortly.
Thanks again.
